examples/tqc/utils.py

```diff
@@ -0,0 +1,410 @@
+            funs_to_decorate=["forward", "get_dist"],
+        )
+
+        self.convert_to_functional(
+            qvalue_network,
+            "critic",
+            create_target_params=True  # Create a target critic network
+        )
+
+        self.device = device
+        self.log_alpha = torch.tensor([np.log(alpha_init)], requires_grad=True, device=self.device)
+        self.gamma = gamma
+        self.top_quantiles_to_drop = top_quantiles_to_drop
+
+        # Compute target entropy
+        action_spec = getattr(self.actor, "spec", None)
+        if action_spec is None:
+            print("Could not deduce action spec from actor network.")
+        if not isinstance(action_spec, CompositeSpec):
+            action_spec = CompositeSpec({"action": action_spec})
+        action_container_len = len(action_spec.shape)
+        self.target_entropy = -float(action_spec["action"].shape[action_container_len:].numel())
+
+    def forward(self, tensordict: TensorDictBase) -> TensorDictBase:
+        td_next = tensordict["next"]
+        reward = td_next["reward"]
+        not_done = tensordict["done"].logical_not()
+
+        alpha = torch.exp(self.log_alpha)
+
+        # Q-loss
+        with torch.no_grad():
+            # get policy action
+            td_next = self.actor(td_next, params=self.actor_params)
+            td_next = self.critic(td_next, params=self.target_critic_params)
+
+            next_log_pi = td_next["sample_log_prob"]
+            next_log_pi = torch.unsqueeze(next_log_pi, dim=-1)
+
+            # compute and cut quantiles at the next state
+            next_z = td_next["state_action_value"]
+            sorted_z, _ = torch.sort(next_z.reshape(next_z.shape[0], -1))
+            sorted_z_part = sorted_z[:, :-self.top_quantiles_to_drop]
+
+            # compute target
+            target = reward + not_done * self.gamma * (sorted_z_part - alpha * next_log_pi)
+
+        td_cur = tensordict
+        td_cur = self.critic(td_cur, params=self.critic_params)
+        cur_z = td_cur["state_action_value"]
+        critic_loss = quantile_huber_loss_f(cur_z, target)
+
+        # --- Policy and alpha loss ---
+        td_new = tensordict
+        td_new = self.actor(td_new, params=self.actor_params)
+        td_new = self.critic(td_new, params=self.critic_params)
+        new_log_pi = td_new["sample_log_prob"]
+        alpha_loss = -self.log_alpha * (new_log_pi + self.target_entropy).detach().mean()
+        actor_loss = (alpha * new_log_pi - td_new["state_action_value"].mean(2).mean(1, keepdim=True)).mean()
+
+        # --- Entropy ---
+        with set_exploration_type(ExplorationType.RANDOM):
+            dist = self.actor.get_dist(
+                tensordict,
+                params=self.actor_params,
+            )
+            a_reparm = dist.rsample()
+        log_prob = dist.log_prob(a_reparm).detach()
+        entropy = -log_prob.mean()
+
+        return TensorDict(
+            {
+                "loss_critic": critic_loss,
```

Comment:
Going further we'd like to separate the methods to compute the actor and critic loss. See how TD3 is coded currently.
Would it make sense here?